keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -12,124 +11,477 @@ approvers:
+> E.g. wildcard "*.foo.com" matches "bar.foo.com" because they share an equal
+> suffix "foo.com".
+
+### Healthchecks
+
+The current spec does not have any provisions to customize appropriate
+healtchecks for referenced backends. Many users already have a healthcheck
+URLthat is lightweight and different from the HTTP root (e.g. `/`).
+
+One option that has been explored is to infer the healthcheck URL from the
+Readiness probes on the Pods of the Service. This method has proven to be quite
+unstable: Every Pod in a Service can have a different Readiness probe definition
+and it's not clear which one should be used. Furthermore, the behavior is quite
+implicit and creates action-at-a-distance relationship between the Ingress and
+Pod resources.
+
+#### Proposal
+
+Add the following fields to `IngressBackend`:
+
+```golang
+type IngressBackend struct {
+  ...
+  // Healthcheck defines custom healthcheck for this backend.
+  // +optional
+  Healthcheck *IngressBackendHealthcheck
+}
+
+type IngressBackendHealthcheck struct {
+  // HTTP defines healthchecks using the HTTP protocol.
+  HTTP *IngressBackendHTTPHealthcheck
+}
+
+// IngressBackendHTTPHealthcheck is a healthcheck using the HTTP protocol.
+type IngressBackendHTTPHealthcheck struct {
+  // Host header to send when healthchecking. If empty, the host header will be
+  // implementation specific.
+  Host string
+  // Path to use for the HTTP healthcheck. If empty, the root '/' path will be
+  // used for healthchecking.
+  Path string
+  // TimeoutSeconds for the healthcheck. Failure to respond with a success code
+  // within TimeoutSeconds will be counted towards the FailureThreshold.
+  TimeoutSeconds int
+  // FailureThreshold is the number of consecutive failures necesseary to
+  // indicate a backend failure.
+  FailureThreshold int
+}
+```
+
+If `Healthcheck` is nil, then the implementation default healthcheck will be
+configured, healthchecking the root `/` path. If `Healthcheck` is specfied,
+then the backend health will be checked using the parameters listed above. 
+
+### Status
+
+!!! TODO
+
+### Ingress class
+
+The `kubernetes.io/ingress.class` annotation is required for interoperation
+between various Ingress providers. As support for this annotation is universal,
+this concept should be promoted to an actual field.
+
+#### Proposal
+
+Promoting the annotation as is as an opaque string the most direct path, but
+precludes any future enhancements to the concept.
+
+An alternative is to create a new resource `IngressClass` to take its place.
+This resource will serve a couple of purposes:
+
+* Define the set of valid classes available to the user. Gives operators control
+  over allowed classes.
+* Allow us to evolve the API to express concepts such a levels of service
+  associated with a given Ingress controller.
+
+Add a field to `ingress.spec`:
+```golang
+type IngressSpec struct {
+  ...
+  // Class references an IngressClass resource in kube-system. This is used
```

Comment:
I think this IngressClass should be a cluster scope resource instead of namespaced resource in kube-system

Reply:
Agree, it should be like storage class